src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -0,0 +1,235 @@
+  /** Holds the tuples that are ready for release. */
+  private transient TupleBatchBuffer buffer;
+  /** Which line of the file the scanner is currently on. */
+  private long lineNumber = 0;
+
+  private final boolean isLastWorker;
+  private final long fileSize;
+
+  private long byteOverlap = 10;
+  private long startByteRange;
+  private long endByteRange;
+  long workerID;
+
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /**
+   * The logger for debug, trace, etc. messages in this class.
+   */
+  private static final org.slf4j.Logger LOGGER = org.slf4j.LoggerFactory.getLogger(CSVFileScanFragment.class);
+
+  public CSVFileScanFragment(final String filename, final Schema schema, final int workerID, final int totalWorkers) {
+    this(filename, schema, workerID, totalWorkers, null, null, null, null);
+  }
+
+  public CSVFileScanFragment(final DataSource source, final Schema schema, final int workerID, final int totalWorkers) {
+    this(source, schema, workerID, totalWorkers, null, null, null, null);
+  }
+
+  public CSVFileScanFragment(final String filename, final Schema schema, final int workerID, final int totalWorkers,
+      final Character delimiter) {
+    this(new FileSource(filename), schema, workerID, totalWorkers, delimiter, null, null, null);
+  }
+
+  public CSVFileScanFragment(final DataSource source, final Schema schema, final int workerID, final int totalWorkers,
+      final Character delimiter) {
+    this(source, schema, workerID, totalWorkers, delimiter, null, null, null);
+  }
+
+  public CSVFileScanFragment(final String filename, final Schema schema, final int workerID, final int totalWorkers,
+      @Nullable final Character delimiter, @Nullable final Character quote, @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this(new FileSource(filename), schema, workerID, totalWorkers, delimiter, quote, escape, numberOfSkippedLines);
+  }
+
+  public CSVFileScanFragment(final DataSource source, final Schema schema, final int workerID, final int totalWorkers,
+      @Nullable final Character delimiter, @Nullable final Character quote, @Nullable final Character escape,
+      @Nullable final Integer numberOfSkippedLines) {
+    this.source = (AmazonS3Source) Preconditions.checkNotNull(source, "source");
+    this.schema = Preconditions.checkNotNull(schema, "schema");
+
+    this.delimiter = MoreObjects.firstNonNull(delimiter, CSVFormat.DEFAULT.getDelimiter());
+    this.quote = MoreObjects.firstNonNull(quote, CSVFormat.DEFAULT.getQuoteCharacter());
+    this.escape = escape != null ? escape : CSVFormat.DEFAULT.getEscapeCharacter();
+    this.numberOfSkippedLines = MoreObjects.firstNonNull(numberOfSkippedLines, 0);
+    this.workerID = workerID;
+
+    isLastWorker = workerID == totalWorkers;
+    fileSize = ((AmazonS3Source) source).getFileSize();
+
+    if (fileSize < 10 && workerID == 1) {
```

Comment:
Maybe the coordinator could just compute a set of workers and the total object size, include those parameters with the local operator sent to each worker, and let each worker compute its own range? Of course, computing the ranges on the coordinator would be trivial, so we could just include the per-worker range in the serialized local operator for simplicity. Workers would still have to extend their ranges for incomplete tuples, of course.